File: 38-Count and Say-外观数列/38-Count and Say-外观数列.java
class Solution {
    public String countAndSay(int n) {
        // 初始化边界条件
        String ans = "1";
        // 从第2项到第n项进行描述
        for (int i = 2; i <= n; i++) {
            //记录当前描述项
            String cur = "";
            // 前一项长度
            int m = ans.length();
            for (int j = 0; j < m; ) {
                int k = j+1;
                // 读取前一项相同字符的长度
                while (k < m && ans.charAt(j) == ans.charAt(k)){
                    k++;
                }
                // 相同字符的数量
                int cnt = k - j;
                // 写入描述信息
                cur += cnt+""+ans.charAt(j);
                j = k;
            }
            // 更新数组
            ans = cur;
        }
        // 返回
        return ans;
    }
}

File: 38-Count and Say-外观数列/38-Count and Say-外观数列.js
// var countAndSay = function(n) {
//     let prev = "1"
//     for(let i = 1; i < n; i++){
//         let temp = ""
//         let prevString = String(prev)
//         let lastCount = ""
//         let lastNumber = ""
//         for(let j = 0; j < prevString.length; j++){
//             if(j === prevString.length - 1){
//                 if(lastNumber !== prevString[j]){
//                     temp +=String(lastCount) + String(lastNumber) + String(1) + String(prevString[j])
//                 }else{
//                     temp += String(lastCount + 1) + String(lastNumber)
//                 }
//                 continue
//             }
//             if(lastNumber !== prevString[j]){
//                 temp += String(lastCount) + String(lastNumber)
//                 lastCount = 1
//                 lastNumber = prevString[j]
//             }else{
//                 lastCount++
//                 lastNumber = prevString[j]
//             }
//         }
//         prev = temp
//     }
//     return prev
// }

/**
 * @param {number} n
 * @return {string}
 */
 var countAndSay = function(n) {
    let prev = "1"
    for(let i = 1; i < n; i++){
        let prevString = String(prev)
        let cur = ''
        debugger
        let len = prevString.length
        for(let j = 0; j < len; ){
            let k = j + 1
            while(k < len && prevString[j] === prevString[k]){
                k++
            }
            cur += String(k - j) + prevString[j]
            j = k
        }
        prev = cur
    }
    return prev
}
